File: basic_dsa_six.java
import java.util.*;

/**
 * Basic DSA Problems - Two More Fundamental Questions
 * Simple and essential DSA problems for beginners
 */
public class basic_dsa_six {
    
    // Problem 1: Fibonacci Sequence
    // Generate Fibonacci sequence using both iterative and recursive approaches
    
    /**
     * Generate first n Fibonacci numbers using iterative approach
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     */
    public static List<Integer> fibonacciIterative(int n) {
        List<Integer> fib = new ArrayList<>();
        if (n <= 0) {
            return fib;
        }
        if (n == 1) {
            fib.add(0);
            return fib;
        }
        if (n == 2) {
            fib.add(0);
            fib.add(1);
            return fib;
        }
        
        fib.add(0);
        fib.add(1);
        for (int i = 2; i < n; i++) {
            fib.add(fib.get(i - 1) + fib.get(i - 2));
        }
        
        return fib;
    }
    
    /**
     * Get nth Fibonacci number using recursive approach
     * Time Complexity: O(2^n) - exponential
     * Space Complexity: O(n) due to recursion stack
     */
    public static Integer fibonacciRecursive(int n) {
        if (n <= 0) {
            return null;
        }
        if (n == 1) {
            return 0;
        }
        if (n == 2) {
            return 1;
        }
        
        return fibonacciRecursive(n - 1) + fibonacciRecursive(n - 2);
    }
    
    /**
     * Get nth Fibonacci number using iterative approach (optimized)
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public static Integer fibonacciNthIterative(int n) {
        if (n <= 0) {
            return null;
        }
        if (n == 1) {
            return 0;
        }
        if (n == 2) {
            return 1;
        }
        
        int a = 0, b = 1;
        for (int i = 2; i < n; i++) {
            int temp = a + b;
            a = b;
            b = temp;
        }
        
        return b;
    }
    
    // Problem 2: Check if a Number is Palindrome
    // Determine whether a number reads the same forwards and backwards
    
    /**
     * Check if a number is palindrome
     * Time Complexity: O(log n) where log is base 10
     * Space Complexity: O(1)
     */
    public static boolean isPalindromeNumber(int n) {
        if (n < 0) {
            return false; // Negative numbers are not palindromes
        }
        if (n < 10) {
            return true; // Single digit numbers are palindromes
        }
        
        int original = n;
        int reversedNum = 0;
        
        while (n > 0) {
            reversedNum = reversedNum * 10 + n % 10;
            n /= 10;
        }
        
        return original == reversedNum;
    }
    
    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem 1: Fibonacci Sequence");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - First 10 Fibonacci numbers
        int n1 = 10;
        List<Integer> fib1 = fibonacciIterative(n1);
        System.out.println("First " + n1 + " Fibonacci numbers (Iterative): " + fib1);
        
        // Test Case 2 - First 5 Fibonacci numbers
        int n2 = 5;
        List<Integer> fib2 = fibonacciIterative(n2);
        System.out.println("\nFirst " + n2 + " Fibonacci numbers (Iterative): " + fib2);
        
        // Test Case 3 - 7th Fibonacci number (iterative optimized)
        int n3 = 7;
        Integer fibNth1 = fibonacciNthIterative(n3);
        System.out.println("\n" + n3 + "th Fibonacci number (Iterative): " + fibNth1);
        
        // Test Case 4 - 7th Fibonacci number (recursive)
        Integer fibNth2 = fibonacciRecursive(n3);
        System.out.println(n3 + "th Fibonacci number (Recursive): " + fibNth2);
        
        // Test Case 5 - Edge cases
        System.out.println("\n1st Fibonacci number: " + fibonacciNthIterative(1));
        System.out.println("2nd Fibonacci number: " + fibonacciNthIterative(2));
        System.out.println("3rd Fibonacci number: " + fibonacciNthIterative(3));
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Problem 2: Check if a Number is Palindrome");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - Palindrome number
        int num1 = 121;
        boolean result1 = isPalindromeNumber(num1);
        System.out.println("Number: " + num1);
        System.out.println("Is Palindrome: " + result1);
        
        // Test Case 2 - Non-palindrome number
        int num2 = 123;
        boolean result2 = isPalindromeNumber(num2);
        System.out.println("\nNumber: " + num2);
        System.out.println("Is Palindrome: " + result2);
        
        // Test Case 3 - Single digit
        int num3 = 5;
        boolean result3 = isPalindromeNumber(num3);
        System.out.println("\nNumber: " + num3);
        System.out.println("Is Palindrome: " + result3);
        
        // Test Case 4 - Palindrome with even digits
        int num4 = 1221;
        boolean result4 = isPalindromeNumber(num4);
        System.out.println("\nNumber: " + num4);
        System.out.println("Is Palindrome: " + result4);
        
        // Test Case 5 - Large palindrome
        int num5 = 12321;
        boolean result5 = isPalindromeNumber(num5);
        System.out.println("\nNumber: " + num5);
        System.out.println("Is Palindrome: " + result5);
        
        // Test Case 6 - Negative number
        int num6 = -121;
        boolean result6 = isPalindromeNumber(num6);
        System.out.println("\nNumber: " + num6);
        System.out.println("Is Palindrome: " + result6);
        
        // Test Case 7 - Number with zeros
        int num7 = 1001;
        boolean result7 = isPalindromeNumber(num7);
        System.out.println("\nNumber: " + num7);
        System.out.println("Is Palindrome: " + result7);
        
        // Test Case 8 - Non-palindrome
        int num8 = 12345;
        boolean result8 = isPalindromeNumber(num8);
        System.out.println("\nNumber: " + num8);
        System.out.println("Is Palindrome: " + result8);
    }
}

File: basic_dsa_six.py
"""
Basic DSA Problems - Two More Fundamental Questions
Simple and essential DSA problems for beginners
"""


# Problem 1: Fibonacci Sequence
# Generate Fibonacci sequence using both iterative and recursive approaches

def fibonacci_iterative(n):
    """
    Generate first n Fibonacci numbers using iterative approach
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if n <= 0:
        return []
    if n == 1:
        return [0]
    if n == 2:
        return [0, 1]
    
    fib = [0, 1]
    for i in range(2, n):
        fib.append(fib[i - 1] + fib[i - 2])
    
    return fib


def fibonacci_recursive(n):
    """
    Get nth Fibonacci number using recursive approach
    Time Complexity: O(2^n) - exponential (can be optimized with memoization)
    Space Complexity: O(n) due to recursion stack
    """
    if n <= 0:
        return None
    if n == 1:
        return 0
    if n == 2:
        return 1
    
    return fibonacci_recursive(n - 1) + fibonacci_recursive(n - 2)


def fibonacci_nth_iterative(n):
    """
    Get nth Fibonacci number using iterative approach (optimized)
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if n <= 0:
        return None
    if n == 1:
        return 0
    if n == 2:
        return 1
    
    a, b = 0, 1
    for _ in range(2, n):
        a, b = b, a + b
    
    return b


# Problem 2: Check if a Number is Palindrome
# Determine whether a number reads the same forwards and backwards

def is_palindrome_number(n):
    """
    Check if a number is palindrome
    Time Complexity: O(log n) where log is base 10
    Space Complexity: O(1)
    """
    if n < 0:
        return False  # Negative numbers are not palindromes
    if n < 10:
        return True  # Single digit numbers are palindromes
    
    original = n
    reversed_num = 0
    
    while n > 0:
        reversed_num = reversed_num * 10 + n % 10
        n //= 10
    
    return original == reversed_num


# Test cases
if __name__ == "__main__":
    print("=" * 60)
    print("Problem 1: Fibonacci Sequence")
    print("=" * 60)
    
    # Test Case 1 - First 10 Fibonacci numbers
    n1 = 10
    fib1 = fibonacci_iterative(n1)
    print(f"First {n1} Fibonacci numbers (Iterative): {fib1}")
    
    # Test Case 2 - First 5 Fibonacci numbers
    n2 = 5
    fib2 = fibonacci_iterative(n2)
    print(f"\nFirst {n2} Fibonacci numbers (Iterative): {fib2}")
    
    # Test Case 3 - 7th Fibonacci number (iterative optimized)
    n3 = 7
    fib_nth1 = fibonacci_nth_iterative(n3)
    print(f"\n{n3}th Fibonacci number (Iterative): {fib_nth1}")
    
    # Test Case 4 - 7th Fibonacci number (recursive)
    fib_nth2 = fibonacci_recursive(n3)
    print(f"{n3}th Fibonacci number (Recursive): {fib_nth2}")
    
    # Test Case 5 - Edge cases
    print(f"\n1st Fibonacci number: {fibonacci_nth_iterative(1)}")
    print(f"2nd Fibonacci number: {fibonacci_nth_iterative(2)}")
    print(f"3rd Fibonacci number: {fibonacci_nth_iterative(3)}")
    
    print("\n" + "=" * 60)
    print("Problem 2: Check if a Number is Palindrome")
    print("=" * 60)
    
    # Test Case 1 - Palindrome number
    num1 = 121
    result1 = is_palindrome_number(num1)
    print(f"Number: {num1}")
    print(f"Is Palindrome: {result1}")
    
    # Test Case 2 - Non-palindrome number
    num2 = 123
    result2 = is_palindrome_number(num2)
    print(f"\nNumber: {num2}")
    print(f"Is Palindrome: {result2}")
    
    # Test Case 3 - Single digit
    num3 = 5
    result3 = is_palindrome_number(num3)
    print(f"\nNumber: {num3}")
    print(f"Is Palindrome: {result3}")
    
    # Test Case 4 - Palindrome with even digits
    num4 = 1221
    result4 = is_palindrome_number(num4)
    print(f"\nNumber: {num4}")
    print(f"Is Palindrome: {result4}")
    
    # Test Case 5 - Large palindrome
    num5 = 12321
    result5 = is_palindrome_number(num5)
    print(f"\nNumber: {num5}")
    print(f"Is Palindrome: {result5}")
    
    # Test Case 6 - Negative number
    num6 = -121
    result6 = is_palindrome_number(num6)
    print(f"\nNumber: {num6}")
    print(f"Is Palindrome: {result6}")
    
    # Test Case 7 - Number with zeros
    num7 = 1001
    result7 = is_palindrome_number(num7)
    print(f"\nNumber: {num7}")
    print(f"Is Palindrome: {result7}")
    
    # Test Case 8 - Non-palindrome
    num8 = 12345
    result8 = is_palindrome_number(num8)
    print(f"\nNumber: {num8}")
    print(f"Is Palindrome: {result8}")
